src/librustc_driver/driver.rs

```diff
@@ -1650,17 +1649,14 @@ pub fn build_output_filenames(
             // "-" as input file will cause the parser to read from stdin so we
             // have to make up a name
             // We want to toss everything after the final '.'
-            let dirpath = match *odir {
-                Some(ref d) => d.clone(),
-                None => PathBuf::new(),
-            };
+            let dirpath = (*odir).as_ref().cloned().unwrap_or_else(|| PathBuf::new());
 
             // If a crate name is present, we use it as the link name
             let stem = sess.opts
                 .crate_name
                 .clone()
                 .or_else(|| attr::find_crate_name(attrs).map(|n| n.to_string()))
-                .unwrap_or(input.filestem());
+                .unwrap_or_else(|| input.filestem());
```

Comment:
I _think_ this should work: `.unwrap_or_else(input.filestem);`

Reply:
That would be cool, but unfortunately the compiler recognizes this as `E0615`.

Reply:
That's annoying but can see how it happens. Thanks for checking.